dsc_lib/src/discovery/command_discovery.rs

```diff
@@ -78,74 +62,241 @@ impl CommandDiscovery {
+                        // In case of "resource list" operation - print failure from adapter as warning
+                        warn!("Adapter failed to list resources with exit code {exit_code}: {stderr}");
+                    }
+
+                    for line in stdout.lines() {
+                        match serde_json::from_str::<DscResource>(line){
+                            Result::Ok(resource) => {
+                                if resource.require_adapter.is_none() {
+                                    warn!("{}", DscError::MissingRequires(adapter.0.clone(), resource.type_name.clone()).to_string());
+                                    continue;
+                                }
+                                if type_regex.is_match(&resource.type_name) {
+                                    resources.insert(resource.type_name.to_lowercase(), resource);
+                                    adapter_resources_count += 1;
+                                }
+                            },
+                            Result::Err(err) => {
+                                warn!("Failed to parse resource: {line} -> {err}");
+                                continue;
                             }
-                            else if remaining_required_resource_types.contains(&resource.type_name.to_lowercase())
-                            {
-                                remaining_required_resource_types.retain(|x| *x != resource.type_name.to_lowercase());
-                                debug!("Found {} in {}", &resource.type_name, path.display());
-                                Span::current().pb_inc(1);
-                                resources.insert(resource.type_name.to_lowercase(), resource);
-                                if remaining_required_resource_types.is_empty()
+                        };
+                    }
+
+                    debug!("Adapter '{}' listed {} matching resources", adapter_type_name, adapter_resources_count);
+                }
+            }
+        }
+        Ok(resources)
+    }
+
+    #[allow(clippy::too_many_lines)]
+    fn discover_resources(&mut self, required_resource_types: &[String]) -> Result<BTreeMap<String, DscResource>, DscError>
+    {
+        debug!("Searching for resources: {:?}", required_resource_types);
+
+        let pb_span = warn_span!("");
+        pb_span.pb_set_style(&ProgressStyle::with_template(
+            "{spinner:.green} [{elapsed_precise:.cyan}] [{bar:40.cyan/blue}] {pos:>7}/{len:7} {msg:.yellow}"
+        )?);
+        pb_span.pb_set_message("Searching for resources");
+        let _ = pb_span.enter();
+
+        let mut resources: BTreeMap<String, DscResource> = BTreeMap::new();
+        let mut adapter_resources: BTreeMap<String, DscResource> = BTreeMap::new();
+        let mut remaining_required_resource_types = required_resource_types.to_owned();
+
+        if let Ok(paths) = CommandDiscovery::get_resource_paths() {
```

Comment:
Seems like there's some redundancy with `list_availble_resource()` above?  Should refactor and have one depend on the other